File: backend/lib/core/rate_limit.py
"""
Hàng đợi tuần tự theo key, có khoảng cách tối thiểu giữa hai lần chạy.

TikTok bắt đầu trả 40100 "too many requests" sau khoảng năm lần gọi liên tiếp, và một
nguồn bị chặn tốn kém hơn nhiều so với một nguồn chạy chậm — nên mọi request ra ngoài
đều đi qua đây thay vì trông chờ nơi gọi tự giữ nhịp.
"""

from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass, field
from typing import Awaitable, Callable, TypeVar

T = TypeVar("T")


@dataclass
class _Queue:
    lock: asyncio.Lock = field(default_factory=asyncio.Lock)
    last_run_at: float = 0.0


_queues: dict[str, _Queue] = {}


def _now_ms() -> float:
    return time.monotonic() * 1000


async def schedule(key: str, min_interval_ms: float, task: Callable[[], Awaitable[T]]) -> T:
    """
    Chạy `task` khi hàng đợi của `key` rảnh và đã qua ít nhất `min_interval_ms` kể từ task
    trước trên cùng key. Hai task cùng key không bao giờ chạy song song.

    Khoá được giữ suốt thời gian chờ lẫn thời gian chạy, nên một task lỗi vẫn nhả khoá và
    không làm nghẽn cả hàng đợi — `finally` bên dưới lo phần đó.
    """
    queue = _queues.get(key)
    if queue is None:
        queue = _Queue()
        _queues[key] = queue

    async with queue.lock:
        wait_for = queue.last_run_at + min_interval_ms - _now_ms()
        if wait_for > 0:
            await asyncio.sleep(wait_for / 1000)
        try:
            return await task()
        finally:
            queue.last_run_at = _now_ms()


def queue_delay(key: str, min_interval_ms: float) -> float:
    """Số mili-giây còn phải chờ trước khi task tiếp theo trên `key` được chạy."""
    queue = _queues.get(key)
    if queue is None:
        return 0
    return max(0.0, queue.last_run_at + min_interval_ms - _now_ms())

File: lib/core/rate-limit.ts
/**
 * Hàng đợi tuần tự theo key, có khoảng cách tối thiểu giữa hai lần chạy.
 *
 * TikTok bắt đầu trả 40100 "too many requests" sau khoảng năm lần gọi liên tiếp, và một
 * nguồn bị chặn tốn kém hơn nhiều so với một nguồn chạy chậm — nên mọi request ra ngoài
 * đều đi qua đây thay vì trông chờ nơi gọi tự giữ nhịp.
 */

type Task<T> = () => Promise<T>

type Queue = {
  chain: Promise<unknown>
  lastRunAt: number
}

const queues = new Map<string, Queue>()

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms))

/**
 * Chạy `task` khi hàng đợi của `key` rảnh và đã qua ít nhất `minIntervalMs` kể từ task
 * trước trên cùng key. Hai task cùng key không bao giờ chạy song song.
 */
export function schedule<T>(key: string, minIntervalMs: number, task: Task<T>): Promise<T> {
  const queue = queues.get(key) ?? { chain: Promise.resolve(), lastRunAt: 0 }
  queues.set(key, queue)

  const run = queue.chain.then(async () => {
    const waitFor = queue.lastRunAt + minIntervalMs - Date.now()
    if (waitFor > 0) await sleep(waitFor)
    try {
      return await task()
    } finally {
      queue.lastRunAt = Date.now()
    }
  })

  // Giữ chuỗi sống ngay cả khi một task lỗi, để một lần hỏng không làm nghẽn cả hàng đợi.
  queue.chain = run.then(
    () => undefined,
    () => undefined,
  )
  return run as Promise<T>
}

/** Số mili-giây còn phải chờ trước khi task tiếp theo trên `key` được chạy. */
export function queueDelay(key: string, minIntervalMs: number): number {
  const queue = queues.get(key)
  if (!queue) return 0
  return Math.max(0, queue.lastRunAt + minIntervalMs - Date.now())
}
